pkg/webreg/webreg.go

```diff
@@ -827,6 +872,81 @@ func setWorkflowGraph(t *template.Template, chains registry.ChainByName, workflo
 		})
 }
 
+type environmentLine struct {
+	Documentation string
+	Default       *string
+	Steps         []string
+}
+
+type environmentData struct {
+	Items map[string]environmentLine
+	Type  string
+}
+
+func getEnvironmentDataItems(worklist []api.TestStep, registryRefs registry.ReferenceByName, registryChains registry.ChainByName) map[string]environmentLine {
+	var data map[string]environmentLine
+
+	add := func(name, documentation, step string, defaultVal *string) {
+		if data == nil {
+			data = map[string]environmentLine{}
+		}
+
+		if _, ok := data[name]; !ok {
+			data[name] = environmentLine{
+				Documentation: documentation,
+				Default:       defaultVal,
+			}
+		}
+
+		line := data[name]
+		line.Steps = append(line.Steps, step)
+		data[name] = line
+	}
+
+	chainWorklist := sets.NewString()
+	seenChains := sets.NewString()
+	dispatch := func(step api.TestStep) {
+		switch {
+		case step.Reference != nil:
+			ref, ok := registryRefs[*step.Reference]
+			if !ok {
+				logrus.WithField("step-name", *step.Reference).Error("failed to resolve step environment, step not found in registry")
+				return
+			}
+			for _, env := range ref.Environment {
+				add(env.Name, env.Documentation, ref.As, env.Default)
+			}
+		case step.Chain != nil:
+			if !seenChains.Has(*step.Chain) {
+				chainWorklist.Insert(*step.Chain)
+			}
+		case step.LiteralTestStep != nil:
+			for _, env := range step.Environment {
+				add(env.Name, env.Documentation, step.As, env.Default)
+			}
+		}
+	}
+
+	for _, step := range worklist {
+		dispatch(step)
+	}
+
+	for len(chainWorklist) > 0 {
+		name, _ := chainWorklist.PopAny()
+		seenChains.Insert(name)
+		chain, ok := registryChains[name]
+		if !ok {
+			logrus.WithField("chain-name", name).Error("failed to resolve chain environment, chain not found in registry")
+			continue
+		}
+		for _, item := range chain.Steps {
```

Comment:
Chains can refer to other chains, too - is there no reusable graph traversal for this? @bbguimaraes

Reply:
I'm not familiar with this module.  `pkg/registry/resolver` never needed one.

But I don't see why we can't enqueue `chain.Steps` inside `dispatch` and do this as a single uniform iteration?